clients/src/main/java/org/apache/kafka/clients/MetadataCache.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients;
+
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * An internal mutable cache of nodes, topics, and partitions in the Kafka cluster. This keeps an up-to-date Cluster
+ * instance which is optimized for read access.
+ */
+class MetadataCache {
+    private final String clusterId;
+    private final List<Node> nodes;
+    private final Set<String> unauthorizedTopics;
+    private final Set<String> invalidTopics;
+    private final Set<String> internalTopics;
+    private final Node controller;
+    private final Map<TopicPartition, PartitionInfoAndEpoch> partitionsByTopicPartition;
+
+    private Cluster clusterInstance;
```

Comment:
Just double-checking if the intent is to keep this in sync with the map above. Now the `Cluster` is only computed on construction.

Reply:
Right, I previously had the logic to recompute the Cluster when the partition was removed, but took that out since it made the partition count inaccurate. Now we are only removing partitions when the topic subscription has changed.

I'll add the Cluster refresh back.

How about I change `MetadataCache#removePartitionInfo(TopicPartition)` to `MetadataCache#removePartitionInfosForTopic(String)` so we only remove an entire topic rather than a single partition. This should avoid the situation where the partition count in Cluster is inaccurate. That would change the behavior to be like:

```
metadata.setTopics(Collections.singletonList("topic-2"));

// now metadata for other topics is unavailable
assertNull(metadata.fetch().partition(tp));
assertNull(metadata.fetch().partitionCountForTopic("topic-1"));
assertFalse(metadata.partitionInfoIfCurrent(tp).isPresent());
```